src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc

```diff
@@ -107,6 +109,33 @@ void SQLiteVersionedObjects::store_versioned_object(
   storage.update(object);
 }
 
+bool SQLiteVersionedObjects::store_versioned_object_if_state(
+    const DBVersionedObject& object, std::vector<ObjectState> allowed_states
+) const {
+  auto storage = conn->get_storage();
+  auto transaction = storage.transaction_guard();
+  storage.update_all(
+      set(c(&DBVersionedObject::object_id) = object.object_id,
+          c(&DBVersionedObject::checksum) = object.checksum,
+          c(&DBVersionedObject::size) = object.size,
+          c(&DBVersionedObject::create_time) = object.create_time,
+          c(&DBVersionedObject::delete_time) = object.delete_time,
+          c(&DBVersionedObject::commit_time) = object.commit_time,
+          c(&DBVersionedObject::mtime) = object.mtime,
+          c(&DBVersionedObject::object_state) = object.object_state,
+          c(&DBVersionedObject::version_id) = object.version_id,
+          c(&DBVersionedObject::etag) = object.etag,
+          c(&DBVersionedObject::attrs) = object.attrs,
+          c(&DBVersionedObject::version_type) = object.version_type),
+      where(
+          is_equal(&DBVersionedObject::id, object.id) and
+          in(&DBVersionedObject::object_state, allowed_states)
+      )
+  );
+  transaction.commit();
```

Comment:
shouldn't `storage.changes()` be called before the `transaction.commit()`?
The examples I found state that in order to user `changes()` a transaction should be created, but it queries changes before returning.

If we have concurrent calls to this function and `changes` is not "inside" the transaction... could we get wrong values?

Reply:
It should not matter, since changes is a call on the sqlite connection returning the number of changes of the last expression.
```
            int changes() {
                auto con = this->get_connection();
                return sqlite3_changes(con.get());
```

The reason there is an explicit transaction is because of sqlite_orm:

"It will print a number of deleted users (rows). But if you call `changes` without a transaction and your database **is located in file** not in RAM the result will be **0** always cause `sqlite_orm` opens and closes connection every time you call a function without a transaction."

That said, I just pushed the change (+ rebase). 
Better be consistent with the sqlite_orm example. Thanks for pointing that out!